### ISSUE TITLE
Difference between shap_values[0] and shap_values[1]?

### ISSUE THREAD

**Issue by Dseal95** [shap/shap#1252]
Hi, 

I am using SHAP to add some feature importance to the end of a classification model on the titanic dataset. I am using a random forest fitted to X_train_df and y_train_df and the output is predicting is an observation survived or not. When I run the shap_values (below) I get a list. The list is a list of two lists each of the two lists containing arrays, each array being (I assume) the SHAP Values for the observation and the number of arrays = the number of training samples. For example, using the following:

shap_values = explainer.shap_values(X_train_df) 

I return shap_values with the following conditions:

shap_values[0][j] where j is in [0,1.....n] where n is number of samples 

and 

shap_values[1][j] where j is in [0,1.....n].

What is the difference in the shap_values[0] and shap_values[1]? apologies if it is a silly Q but I cannot figure it out!

Also, when doing the explainer.expected_value where my explainer object is:

explainer = shap.TreeExplainer(forest_clf_opt_df, feature_perturbation='tree_path_dependent') and the model is forest_clf_opt_df, I get a 2 element array like this:

array([n1, n2]) where n1 and n2 are just numbers. What does this mean? Is it in relation to the shap_values[0] and shap_values[1]?

Lastly, as I am not sure of the above, even though I can plot the following graph, changing the i and j's changes it drastically for each observation and I am not sure what parameters to use. If I was to plot the features each contributing to push the model output from the base value (the average model output over the training dataset we passed) for observation x0 (the first observation), is the following code correct: .

shap.force_plot(explainer.expected_value[0], shap_values[0][0], X_train_df.iloc[0,:])

For this I take the first element of the explainer.expected_value, the first list of shap_values and then the first array of that list and then take the first observation of my training data. It plots as expected but I get confused because If I plot, 

shap.force_plot(explainer.expected_value[0], shap_values[1][0], X_train_df.iloc[0,:])

or 

shap.force_plot(explainer.expected_value[1], shap_values[0][0], X_train_df.iloc[0,:])

I get completely different values for the same observation. 

any guidance here would be really appreciated. 

Plots (in pdf):

[SHAP Values.pdf](https://github.com/slundberg/shap/files/4734689/SHAP.Values.pdf)

**Comment by InterferencePattern** [comment]
The shap_values[0] are explanations with respect to the negative class, while shap_values[1] are explanations with respect to the positive class. If your model predicts a probability, p, for the positive class, the negative class' predicted probability will be 1-p. 

(The explanation for the 2-dimensional expected values is similar.)

If your model generates probabilities of each class, you most likely want index 1, not index 0.

**Comment by Dseal95** [comment]
Hi @jimbudarz, 

Thanks for explaining that. If you are able to, please could you help to confirm the below (it is one more question that I have based on the above)

 If we look at the following two graphs which are the shap.force_plots for the 1st observation (X_train_df[0]) in my instance:
<img width="994" alt="Screenshot 2020-06-12 at 07 36 18" src="https://user-images.githubusercontent.com/63999160/84472638-79a67e80-ac7f-11ea-8d03-1fd4b28956e2.png">

would this explanation be correct:

**Plot1 -** 
Parameters=
explainer.expected_value[0] = the base value w.r.t the negative class 
shap_values[0][0] = the shap value w.r.t to the negative class and 1st observation 
X_train_df.iloc[0,:] = values of the 1st observation from the training dataset

Explanation = The plot show that observation has a model output value of 0.14 w.r.t the negative class. This means that it was predicted a low likelihood of negative class (not survived) which is correct as the observation survived. Based on this 0.14, you see that male = 1 and female = 0 contributed negatively to the 0.14 meaning that they lowered the likelihood that it was a negative class and then the features in the blue contributed positively to the score. Contributing positively means that they were increasing likelihood of the negative class occurring. 

**Plot2-** 
Parameters=
explainer.expected_value[1] = the base value w.r.t the positive class 
shap_values[1][0] = the shap value w.r.t to the positive class and 1st observation 
X_train_df.iloc[0,:] = values of the 1st observation from the training dataset

Explanation = The plot show that observation has a model output value of 0.86 w.r.t the positive class. This means that it was predicted a high likelihood of positive class (survived) which is correct as the observation survived. Based on this 0.86, you see that male = 1 and female = 0 contributed positively to the 0.86 and then the features in the blue contributed negatively to the score. Contributing negatively means that they were decreasing likelihood of the positive class occurring. 


Thanks for your help. 

Dan

**Comment by InterferencePattern** [comment]
Hi @Dseal95 . You're mostly correct, but I believe you're interpreting the colors/Shapley values incorrectly. Since you're interested in the results with respect to the negative class here, I'll try to be really careful about double negatives...

Ignoring the fact that you have _both_ male=1 and female=0 in your dataset (of which you ought to drop one, as they contain the same information), these features actually contributed **positively** to the prediction of the negative class (i.e. these made the passenger more likely to have _died_.) As you can see by the red "higher" arrow, these can be seen as "pushing" or "forcing" the prediction higher, while the blue "lower" arrows are features that force the prediction lower.

I hope this helps.

**Comment by Solwarda** [comment]
Thanks very much! but I  can't understand the relationship between  shap model output and my xgboost binary classification result , my tree model predict the sample with 0.1398022 but shap give me 1.82 result equals model output value add  base value, what the meaning of 1.82 ?  just  as  the png  showed below, 
![image](https://user-images.githubusercontent.com/13435549/85243768-4af58880-b475-11ea-9304-e03aebb8c9fc.png)

**Comment by InterferencePattern** [comment]
Is it possible the explainer is giving your values in log-odds instead of probability?

**Comment by Dseal95** [comment]
@jimbudarz sorry for not getting back sooner, the recent comment reminded me. Your reply was well received and made a lot of sense. I dropped the male feature as suggested too and re ran my script. 

Dan

**Comment by JSVJ** [comment]
> The shap_values[0] are explanations with respect to the negative class, while shap_values[1] are explanations with respect to the positive class. If your model predicts a probability, p, for the positive class, the negative class' predicted probability will be 1-p.
> 
> (The explanation for the 2-dimensional expected values is similar.)
> 
> If your model generates probabilities of each class, you most likely want index 1, not index 0.

Could you please tell me how are you telling shap_values[0] is negative class ? Any idea how to verify it manually?

**Comment by InterferencePattern** [comment]
@JSVJ
You can verify this in scikit-learn models by looking at the model object's `.classes_` attribute. This will contain an array of the model's classes, in order. The model's `predict_proba()` method will produce arrays of this width, with a probability predicted for each class.

**Comment by JSVJ** [comment]
> @JSVJ
> You can verify this in scikit-learn models by looking at the model object's `.classes_` attribute. This will contain an array of the model's classes, in order. The model's `predict_proba()` method will produce arrays of this width, with a probability predicted for each class.

Thank you very much. I understand now

**Comment by jane-delaney** [comment]
How does this work when you are looking at the log loss shap values? Should you take the values for the positive class where the true label is positive and for the negative class where it is negative?

**Comment by Yuri-Nassar** [comment]
> The shap_values[0] are explanations with respect to the negative class, while shap_values[1] are explanations with respect to the positive class. If your model predicts a probability, p, for the positive class, the negative class' predicted probability will be 1-p.
> 
> (The explanation for the 2-dimensional expected values is similar.)
> 
> If your model generates probabilities of each class, you most likely want index 1, not index 0.

Hi guys.

I'm using a sample dataset from X_train (binary problem) as background passed into `shap.Explainer(model, background)` . My `explainer.expected_value` is not an array, but a float. I'm learning about SHAP, so I don't know if I lost some points on how to use it. I saw other examples like yours to get info regarding the class, but I got errors since my result does not have, apparently, the class on shap. Tkx for any help.

**Comment by Dseal95** [comment]
Hi @Yuri-Nassar, 

Would you be able to explain a little more on the problem? Is it that you have a binary classifier, predicting 0 or 1 and that you have a typical dataset that you are using to train and test this classifier? 

Thanks,
Dan

**Comment by Yuri-Nassar** [comment]
> Hi @Yuri-Nassar,
> 
> Would you be able to explain a little more on the problem? Is it that you have a binary classifier, predicting 0 or 1 and that you have a typical dataset that you are using to train and test this classifier?
> 
> Thanks, Dan

Hi @Dseal95 ,

Thank you for your time.
Yes, correct. I have a binary classifier (xgboost), and I was trying to reproduce the examples of SHAP values for a given class (target). I trained the xgboost (model) with a `X_train` dataset, call the `explanation = shap.Explainer(model)` and then `shap_values = explanation(X_train)` (either X_test). I also tried `shap.Explainer(model.predict/predict_proba)` hoping to get explanations for both classes, but I didn't. So, I'm wondering to know what I'm misunderstanding. In conclusion, I expected to reproduce something similar to your previous post on Jun 12/2020.
